aptos-move/aptos-native-interface/src/builder.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    context::SafeNativeContext,
+    errors::{SafeNativeError, SafeNativeResult},
+};
+use aptos_gas_schedule::{MiscGasParameters, NativeGasParameters};
+use aptos_types::on_chain_config::{Features, TimedFeatures};
+use move_vm_runtime::native_functions::{NativeContext, NativeFunction};
+use move_vm_types::{
+    loaded_data::runtime_types::Type, natives::function::NativeResult, values::Value,
+};
+use smallvec::SmallVec;
+use std::{collections::VecDeque, sync::Arc};
+
+/// Data shared by all native functions, mostly on-chain configurations.
+#[derive(Debug)]
+struct SharedData {
+    gas_feature_version: u64,
+    native_gas_params: NativeGasParameters,
+    misc_gas_params: MiscGasParameters,
+    timed_features: TimedFeatures,
+    features: Features,
+}
+
+/// Factory object that allows one to build native functions with ease.
+///
+/// This enables native functions to access shared data, and interface with `SafeNativeContext`.
+#[derive(Debug)]
+pub struct SafeNativeBuilder {
+    data: Arc<SharedData>,
+    enable_incremental_gas_charging: bool,
+}
+
+impl SafeNativeBuilder {
+    /// Creates a new safe native builder.
+    ///
+    /// The configurations provided will be accessible by all native functions created later.
+    pub fn new(
+        gas_feature_version: u64,
+        native_gas_params: NativeGasParameters,
+        misc_gas_params: MiscGasParameters,
+        timed_features: TimedFeatures,
+        features: Features,
+    ) -> Self {
+        Self {
+            data: Arc::new(SharedData {
+                gas_feature_version,
+                native_gas_params,
+                misc_gas_params,
+                timed_features,
+                features,
+            }),
+            enable_incremental_gas_charging: true,
+        }
+    }
+
+    /// Controls the default incremental gas charging behavior of the natives created from this builder.
+    ///
+    /// Default: enabled.
+    pub fn set_incremental_gas_charging(&mut self, enable: bool) {
+        self.enable_incremental_gas_charging = enable;
```

Comment:
not related but it's not clear to me what this incremental gas charging means

Reply:
The native function API provided by the Move VM is rather primitive in terms of gas metering: it requires a native functions to calculate the total gas cost itself, and charge it after it returns. This means we are doing work before charging gas, which is not good. 

`SafeNativeContext` was initially made to address this problem by maintaining a counter of the current gas usage, and it aborts early when the accumulative usage has exceeded the remaining balance. This was neat cuz it required minimal changes to the Move VM itself (only need to expose the remaining balance). However, this is still a behavior change, so if we want to make the switch, we'll need to gate it properly.

Reply:
Added a line of comment pointing to `SafeNativeContext::set_incremental_gas_charging`, which explains this in details.